### ISSUE TITLE
Convergence speed difference between pytorch and torch

### ISSUE THREAD

**Issue by Minju-Jung** [SeanNaren/deepspeech.pytorch#61]
I installed both torch and pytoch version of deep speech2.
However, I noticed that torch version converges much faster than pytorch on AN4 dataset.

In pytorch version, I followed default setting of torch version by using the following command:
python train.py --cuda --train_manifest data/an4_train_manifest.csv --val_manifest data/an4_val_manifest.csv --rnn_type rnn --hidden_size 1760 --hidden_layers 7 --noise_prob 0

I can not figure out what makes a big convergence speed difference.

**Comment by SeanNaren** [comment]
It might be the actual RNN, the RNN in the torch lua is the true deepspeech RNN, whereas in this library it's a standard RNN. I'm currently fixing the skip_input implementation in pytorch in #4 that will hopefully allow us to use the same architecture!

**Comment by Minju-Jung** [comment]
Thank you for fast reply!
I didn't know about skip connection in RNN.

Could you kindly let me know what skip connection exactly works in RNN?
NVIDIA documentation is written as "No operation is performed at the input of the first recurrent layer."
Is it mean that "h_{t}=f(x_{t} + U * h_{t-1} +b)" without forward weights?

Also, what's the effect or purpose of skip input?

**Comment by SeanNaren** [comment]
No worries man! Thats exactly it, there is no matrix multiply with the input. What that means is we can apply the linear multiply outside the RNN, and then batch norm. Then we can pass it into the RNN (which is the proper way of applying batch normalization to this input). Hopefully that makes sense! 

I'd keep this ticket open till we prove that this is the reason convergence is slower :)

**Comment by Minju-Jung** [comment]
Thank you very much!
Now, I understand the skip connection and reason of slow-down.

I will wait for your revised version.

**Comment by dlmacedo** [comment]
So we have to wait until #4?

**Comment by SeanNaren** [comment]
@dlmacedo that's correct!

**Comment by dlmacedo** [comment]
Any news about this?

**Comment by SeanNaren** [comment]
My guess is that its due to the RNN, but thats not going to be fixed #4
